cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -96,70 +95,70 @@ __device__ __host__ T clip(T val, T lb, T ub) {
 }
 
 /**
-	         * Calculate the repulsive gradient
-	         */
-template <typename T>
-__device__ __host__ T repulsive_grad(T dist_squared, float gamma,
-                                     UMAPParams params) {
-  T grad_coeff = 2.0 * gamma * params.b;
+ * Calculate the repulsive gradient
+ */
+__device__ __host__ double repulsive_grad(double dist_squared, double gamma,
+                                          UMAPParams params) {
+  double grad_coeff = 2.0 * gamma * params.b;
   grad_coeff /=
     (0.001 + dist_squared) * (params.a * pow(dist_squared, params.b) + 1);
   return grad_coeff;
 }
 
 /**
-	         * Calculate the attractive gradient
-	         */
-template <typename T>
-__device__ __host__ T attractive_grad(T dist_squared, UMAPParams params) {
-  T grad_coeff = -2.0 * params.a * params.b * pow(dist_squared, params.b - 1.0);
+  * Calculate the attractive gradient
+  */
+__device__ __host__ double attractive_grad(double dist_squared,
```

Comment:
Another custom kernel.

Reply:
This is a utility function that happens to be a device function